File: Data Structures & Algorithms/islands-and-treasure/submission-3.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        visited = set()
        q = deque()
        rows, cols = len(grid), len(grid[0])
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 0:
                    q.append([r,c])
                    visited.add((r,c))
        def addroom(r,c):
            if r<0 or c < 0 or r >= rows or c >= cols or (r,c) in visited or grid[r][c] == -1:
                return
            visited.add((r,c))
            q.append([r,c])
        distance = 0
        while q:
            l = len(q)
            for i in range(l):
                r, c = q.popleft()
                grid[r][c] = distance
                addroom(r+1,c)
                addroom(r-1,c)
                addroom(r,c+1)
                addroom(r,c-1)
            distance+=1
        

File: Data Structures & Algorithms/max-area-of-island/submission-3.java
class Solution {
    int[][] dir = {{1,0}, {-1,0},{0,1},{0,-1}};
    public int maxAreaOfIsland(int[][] grid) {
        int rows = grid.length, cols = grid[0].length, area = 0;
        for(int r = 0; r < rows ; r++){
            for(int c = 0; c < cols; c++){
                if(grid[r][c] == 1){
                    area = Math.max(area, dfs(grid, r, c));
                }
            }
        }
        return area;

    }
    private int dfs(int[][] grid, int r, int c){
        if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] == 0){
            return 0;
        }
        grid[r][c] = 0;
        int area = 1;
        for(int[] d : dir){
            area += dfs(grid , r + d[0], c + d[1]);
        }
        return area;
    }
}
